stingray/lightcurve.py

```diff
@@ -789,11 +791,114 @@ def sort(self, reverse=False):
+        """Choose a reasonable chunk length.
+
+        The user specifies a condition on the total counts in each chunk and
+        the minimum number of time bins.
+
+        Other Parameters
+        ----------------
+        min_total_counts : int
+            Minimum number of counts for each chunk
+        min_time_bins : int
+            Minimum number of time bins
+
+        Returns
+        -------
+        chunk_length : float
+            The length of the light curve chunks that satisfies the conditions
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> time = np.arange(150)
+        >>> count = np.zeros_like(time) + 3
+        >>> lc = Lightcurve(time, count)
+        >>> lc.estimate_chunk_length(min_total_counts=10, min_time_bins=3)
+        4.0
+        >>> lc.estimate_chunk_length(min_total_counts=10, min_time_bins=5)
+        5.0
+        >>> count[2:4] = 1
+        >>> lc = Lightcurve(time, count)
+        >>> lc.estimate_chunk_length(min_total_counts=3, min_time_bins=1)
+        4.0
+        """
+
+        rough_estimate = np.ceil(min_total_counts / self.meancounts) * self.dt
+
+        chunk_length = np.max([rough_estimate, min_time_bins * self.dt])
+
+        keep_searching = True
+        while keep_searching:
+            start_times, stop_times, results = \
+                self.analyze_lc_chunks(chunk_length, np.sum)
+            mincounts = np.min(results)
+            if mincounts >= min_total_counts:
+                keep_searching = False
+            else:
+                chunk_length *= np.ceil(min_total_counts / mincounts) * self.dt
+
+        return chunk_length
+
+    def analyze_lc_chunks(self, chunk_length, func, fraction_step=1, **kwargs):
+        """Analyze chunks of the light curve with any function.
+
+        Parameters
+        ----------
+        chunk_length : float
+            Length in seconds of the light curve chunks
+        func : function
+            Function accepting a `Lightcurve` object as single argument, plus
+            possible additional keyword arguments, and returning a number or a
+            tuple - e.g., (result, error) where both result and error are
+            numbers.
+
+        Other parameters
+        ----------------
+        fraction_step : float
+            If the step is not a full chunk_length but less (e.g. a moving window),
+            this indicates the ratio between step step and `chunk_length` (e.g.
+            0.5 means that the window shifts of half chunk_length)
+        kwargs : keyword arguments
+            These additional keyword arguments, if present, they will be passed
+            to `func`
+
+        Returns
+        -------
+        start_times : array
+            Lower time boundaries of all chunks.
+        stop_times : array
+            Higher time boundaries of all chunks.
+        result : array of N elements
+            The result of `func` for each chunk of the light curve
```

Comment:
Yes, GTIs are taken into account here, otherwise it would have been way easier (not that it's complicated).
And yes, np.mean was sufficient. But the typical use case will have the user defining a function, so I defined it, even if it's a little useless :D. I can change the example, if you think it might confuse people.